File: hu/codes/java/chapter_tree/binary_tree_bfs.java
/**
 * File: binary_tree_bfs.java
 * Created Time: 2022-11-25
 * Author: krahets (krahets@163.com)
 */

package chapter_tree;

import utils.*;
import java.util.*;

public class binary_tree_bfs {
    /* Szintenkénti bejárás */
    static List<Integer> levelOrder(TreeNode root) {
        // Sor inicializálása, gyökércsomópont hozzáadása
        Queue<TreeNode> queue = new LinkedList<>();
        queue.add(root);
        // Lista inicializálása a bejárási sorrend mentéséhez
        List<Integer> list = new ArrayList<>();
        while (!queue.isEmpty()) {
            TreeNode node = queue.poll(); // Kivétel a sorból
            list.add(node.val);           // Csomópont értékének mentése
            if (node.left != null)
                queue.offer(node.left);   // Bal gyermekcsomópont sorba rakása
            if (node.right != null)
                queue.offer(node.right);  // Jobb gyermekcsomópont sorba rakása
        }
        return list;
    }

    public static void main(String[] args) {
        /* Bináris fa inicializálása */
        // Egy függvényt használunk a bináris fa közvetlen tömbből való létrehozásához
        TreeNode root = TreeNode.listToTree(Arrays.asList(1, 2, 3, 4, 5, 6, 7));
        System.out.println("\nInitialize binary tree\n");
        PrintUtil.printTree(root);

        /* Szintenkénti bejárás */
        List<Integer> list = levelOrder(root);
        System.out.println("\nLevel-order traversal node print sequence = " + list);
    }
}

File: hu/codes/python/chapter_tree/binary_tree_bfs.py
"""
File: binary_tree_bfs.py
Created Time: 2022-12-20
Author: a16su (lpluls001@gmail.com)
"""

import sys
from pathlib import Path

sys.path.append(str(Path(__file__).parent.parent))
from modules import TreeNode, list_to_tree, print_tree
from collections import deque


def level_order(root: TreeNode | None) -> list[int]:
    """Szintenkénti bejárás"""
    # Sor inicializálása, gyökércsomópont hozzáadása
    queue: deque[TreeNode] = deque()
    queue.append(root)
    # Lista inicializálása a bejárási sorrend mentéséhez
    res = []
    while queue:
        node: TreeNode = queue.popleft()  # Kivétel a sorból
        res.append(node.val)  # Csomópont értékének mentése
        if node.left is not None:
            queue.append(node.left)  # Bal gyermekcsomópont sorba állítása
        if node.right is not None:
            queue.append(node.right)  # Jobb gyermekcsomópont sorba állítása
    return res


"""Fő kód"""
if __name__ == "__main__":
    # Bináris fa inicializálása
    # Tömbből közvetlenül bináris fát generáló függvény használata
    root: TreeNode = list_to_tree(arr=[1, 2, 3, 4, 5, 6, 7])
    print("\nBináris fa inicializálása\n")
    print_tree(root)

    # Szintenkénti bejárás
    res: list[int] = level_order(root)
    print("\nSzintenkénti bejárás csomópont kiírási sorrendje = ", res)
